### DISCUSSION TITLE
Pages Extraction Script

### DISCUSSION THREAD

**Discussion by meteorquake** [sumatrapdfreader/sumatrapdf#4327] [General]
I thought this script I made might be helpful for anyone who wants to save extracts of what they have open in Sumatra directly from Sumatra.
If you have or put pdftk.exe on your system (it often is there from other installed programs) then, the following will add a menu item to extract page ranges (which can be gapped ranges) - you need to change 'D:\Scripts\pdf-extract.vbs' to wherever you put the script -

```
ExternalViewers [
...
	[
		CommandLine = "C:\Windows\System32\wscript.exe" "D:\Scripts\pdf-extract.vbs" "%1"
		Name = Extractor
		Filter = *.pdf
	]
...
}
```

and the script itself as follows which can be saved somewhere, and tweaked however you wish, you can also add it to the explorer "Send To" menu using the same command line as you used just above (for that, run shell:sendto to get to that folder), you will need to change 'C:\Unreg\PdftkBuilderEnhanced\pdftk.exe' to be the location of pdftk.exe on your computer -

```
Set oShell = CreateObject("WScript.Shell")
Set oArgs = WScript.Arguments
Set oFs = CreateObject("Scripting.FileSystemObject")

For iArg = 0 To oArgs.Count - 1
sFile = oArgs(iArg)

sPg = InputBox("Pages (e.g. 7 4-5 2)" & vbCrLf & oFs.GetFileName(sFile), "PDF Page Extractor"): If sPg = "" Then WScript.Quit

sFileN = InputBox("Filename:", "PDF Page Extractor", Left(sFile, Len(sFile) - Len(oFs.GetExtensionName(sFile)) - 1) & "-extract.pdf"): If sFileN = "" Then WScript.Quit

oShell.Run """C:\Unreg\PdftkBuilderEnhanced\pdftk.exe"" """ & sFile & """ cat " & sPg & " output """ & sFileN & """", 3

Next
```

**Comment by GitHubRulesOK** [comment]
Not bad but you can use one cmd line to extract a single page or a whole file without slower VBS as an intermediate. Using 
PDFtotext which is often simpler or use MuPDF or even GhostScript all have simple extraction syntax  for adjusting via CMD "choices"

**Comment by meteorquake** [comment]
Your suggestion is a bit ambiguous - would it ask for page ranges (e.g. 10,17-20) and destination file (e.g. r:\lamium.pdf) ?

(On a side note I would say though that VBS isn't slow; one of its virtues is it is very responsive, and here allows you to customise responses such as output name.)

d

**Reply by GitHubRulesOK** [reply]
There were several replies for similar functions in the past so perhaps this one illustrates one approach **it can be trimmed down a lot** and still offer several functions like OCR for image PDF or output to PNG a selection when it was last updated

It needs a darn good overhaul but as MuPDF 1.20 was last 32 bit it still works for me (on 32 bit win10) and I do have personal 32bit Mutool 1.25

Key line for above comment is `set /p pages="Range = "` which in theory could be written into a CMD "oneliner" but when error checking  and adding descriptions to be used standalone becomes 160 lines.cmd !! https://github.com/GitHubRulesOK/MyNotes/blob/master/AppNotes/SumatraPDF/Addins/ExportPng/ExportPng.cmd

I agree on timing not a lot of delay using VBS or WSH JScript but it is CMD >VBS>wsh.exe >other.EXE whereas CMD >other.exe can sometimes avoid some bigger temporary scripting. But I do use double-click vbs/Jscript on occasion where syntax is better for file building

**Comment by meteorquake** [comment]
I should add that I updated that script a little since, as follows, it now allows for options and handles a few extra situations to my needs, so *L at the end won't launch, lack of a path will prefix with r:\ path, pdf will be added if absent, that allows for better handling of just typing in a pathless title; I also have a few name intercepts which I've removed from this -

```
Option Explicit

Dim oShell, oArgs, oFs, iArg, sFile, sPg, sFileN, sNme,sOpt,i,sCmd

Set oShell = CreateObject("WScript.Shell")
Set oArgs = WScript.Arguments
Set oFs = CreateObject("Scripting.FileSystemObject")

sPg="":sOpt=""

For iArg = 0 To oArgs.Count - 1
sFile = oArgs(iArg)
If Left(sFile,1)="#" Then
sPg=Mid(sFile,2)
Else
sNme=oFs.GetFileName(sFile)

sPg = Replace(InputBox("Pages (e.g. 7,4-5,2 (or spaced))" & vbCrLf & sNme, "PDF Page Extractor",sPg),","," "): If sPg = "" Then WScript.Quit
sFileN = InputBox("Filename*opts caseless:"&vbCrLf &"(L)aunch NOT"&vbCrLf &"Pathless prefixes R:\", "PDF Page Extractor", Left(sFile, Len(sFile) - Len(oFs.GetExtensionName(sFile)) - 1) & "-extract.pdf"): If sFileN = "" Then WScript.Quit
i=Instr(1,sFileN,"*",0):If i Then sOpt=UCase(Mid(sFileN,i+1)):sFileN=Left(sFileN,i-1)
If Right(sFileN,4)<>".pdf" Then sFileN=sFileN &".pdf"
If Instr(1,sFileN,"\",0)=0 Then sFileN="r:\" & sFileN

sCmd = """C:\Program Files\PdftkBuilderEnhanced\pdftk.exe"" """ & sFile & """ cat " & sPg & " output """ & sFileN & """"
oShell.Run sCmd, 3,-1
If Instr(1,sOpt,"L",0)=0 Then oShell.Run """"&sFileN &"""", 3
End If

Next

```

**Reply by GitHubRulesOK** [reply]
the scripting is great but I strongly recommend that to use the only modern tool for reducing pages you switch the tool to coherent `cpdf "mutool clean - MuPDF 1.27.1.pdf" [insert pages range here] -process-struct-trees -o less-pages.pdf` I spent ages testing mupdf page reduction hence the input instructions page and it failed to keep collateral objects like named destinations or file structure unless it ported the whole tree into the reduced pages and links suffered so were removed en-block!
try
```
sCmd = """C:\Program Files\cpdf\cpdf.exe"" """ & sFile & """ " & sPg & " -process-struct-trees -o """ & sFileN & """"
oShell.Run sCmd, 3, -1

```

**Comment by meteorquake** [comment]
I'm not sure what system pdftk.exe uses but I've used it thousands of times and not once had any issue arising unless the file was a protected one. At any rate definitely people should treat the script as a template and adjust as they need, I'm sure they would do so anyway!

**Reply by GitHubRulesOK** [reply]
pdftk uses an ancient iText 2005-10-19 (https://sourceforge.net/projects/itextpdf/files/itext-paulo/itext-paulo-155/) and has a reputation for not porting page resources. However it was state of the art in its day and thus still provides a range of good advanced abilities.
cpdf is way newer (20 years better) and handles accessibility, bookmarks, links fonts, etc far better, even better than MuPdf. Not simply repainting page visuals.

**Comment by GitHubRulesOK** [comment]
@meteorquake Just for your info and perhaps critique I have written a PDF page extractor for call with current file or use standalone via sendto or drag and drop.

<img width="561" height="171" alt="Image" src="https://github.com/user-attachments/assets/0cacfa14-d3f3-4c38-bb8e-c27b9aafc14b" />

I have uploaded that modified script as https://github.com/GitHubRulesOK/SumatraPDF-Plus/blob/master/Plus/ExportPages.cmd  note cpdf has a wide set of page descriptions like export some portrait pages  in a range or just the landscape ones all in the manual at https://www.coherentpdf.com/cpdfmanual.pdf#page=22 1.3 Input Ranges

It needs a writable folder such as "plus" below the SumatraPDF.exe location so ideally a portable variant.
```
ExternalViewers [
...
	[
		CommandLine = "C:\ location \plus\ExportPages.cmd" "%1"
		Name = PDF e&Xtractor
		Filter = *.pdf
		Key = x
	]
...
]
```
The command as written pauses with consoles for initial use "customisation"
but those are easy to remove after tests so it runs fast without comments.

To see WHY CPDF is best look at an example in https://github.com/sumatrapdfreader/sumatrapdf/discussions/4738#discussioncomment-16097334
